cmd/sanitize-prow-jobs/main.go

```diff
@@ -6,36 +6,120 @@ import (
 	"io/ioutil"
 	"os"
 	"path/filepath"
+	"regexp"
+	"sort"
 	"strings"
 	"sync"
 
+	"github.com/pkg/errors"
 	"github.com/sirupsen/logrus"
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 	prowconfig "k8s.io/test-infra/prow/config"
 	"sigs.k8s.io/yaml"
 )
 
-const defaultCluster = "api.ci"
-
 type options struct {
 	prowJobConfigDir string
+	configPath       string
 
 	help bool
 }
 
+type Config struct {
+	Clusters struct {
+		Default    string `json:"default"`
+		Assignment map[string]struct {
+			// Jobs override both files and groups
+			Jobs []string `json:"jobs"`
+			// Files override files
+			Files  []string `json:"files"`
+			Groups []string `json:"groups"`
+		} `json:"assignment"`
+		Groups    map[string][]string       `json:"groups"`
+		FileREMap map[string]*regexp.Regexp `json:"-"`
+	} `json:"clusters"`
+}
+
+func (config *Config) getClusterForJob(jobName string, path string) string {
+	distribution := config.Clusters.Assignment
+	clusters := make([]string, 0, len(distribution))
+	for cluster := range distribution {
+		clusters = append(clusters, cluster)
+	}
+	sort.Strings(clusters)
+
+	//jobs override files which override groups
+	for _, cluster := range clusters {
+		for _, job := range distribution[cluster].Jobs {
+			if jobName == job {
+				return cluster
+			}
+		}
+	}
+	for _, cluster := range clusters {
```

Comment:
This smells like a bad configuration structure. It should not be this hard to determine the cluster for a job. Why is it even possible to list a job in multiple lists/groups/regex-matches?

Why not:

```yaml
groups:
- name: foo
  jobs:
  - bar
  cluster: zing
raw:
- jobs:
  - whoa
  cluster: there
files:
- paths:
  - ^whatever.yaml$
  cluster: boogie
```

Reply:
Or, validate on config load that a job is not in more than one bucket and take it's presence in a bucket as authoritative -- that way if you find it under `cluster.files` or `cluster.groups` you know it belongs in that cluster no matter what.

Reply:
Said another way, we are building a lot of complexity here to allow `cluster1.groups` to override `cluster2.jobs` which seems silly

Reply:
Done.